### PR TITLE
[lsp] properly rename quoted identifiers

### LINKED ISSUE #1488: [lsp] rename fails for quoted identifiers

### PR DESCRIPTION
**Pull request by JanKoehnlein**
Fixes #1488

### CODE REVIEW

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend [hunk 1]
```diff
@@ -219,6 +222,14 @@ class RenameService2 implements IRenameService2 {
 		return null
 	}
 
+	protected def String getConvertedValue(EObject grammarElement, ILeafNode leaf) {
+		switch (grammarElement) {
+			RuleCall: valueConverterService.toValue(leaf.text, grammarElement.rule.name, leaf).toString()
+			CrossReference: getConvertedValue(grammarElement.terminal, leaf)
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend:228]: shouldt we consult `linkingHelper.getRuleNameFrom(updatable.getCrossReference())` e.g. as in referenceupdater.getQualifiedName

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend [hunk 2]
```diff
@@ -197,11 +200,11 @@ class RenameService2 implements IRenameService2 {
 					if (element !== null && !element.eIsProxy) {
 						val leaf = NodeModelUtils.findLeafNodeAtOffset(rootNode, candidateOffset)
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend:201]: Probably beyond this small bugfix here, but the rename infrastructure appears to be based on the assumption that IDs are comprised of leaf nodes and ignores the fact that `MyId: ID | 'someKeyword';` or `VersionedId: ID '-''(' INT ')';` are valid identifiers, too.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend [hunk 3]
```diff
@@ -197,11 +200,11 @@ class RenameService2 implements IRenameService2 {
...
 						if (leaf !== null && leaf.isIdentifier) {
-							val leafText = NodeModelUtils.getTokenText(leaf)
+							val leafText = getConvertedValue(leaf.grammarElement, leaf)
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend:203]: Given that this is already quite a bit of complexity in the logic, please change the variable name to better reflect what it is now.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend [hunk 4]
```diff
@@ -219,6 +222,14 @@ class RenameService2 implements IRenameService2 {
 		return null
 	}
 
+	protected def String getConvertedValue(EObject grammarElement, ILeafNode leaf) {
+		switch (grammarElement) {
+			RuleCall: valueConverterService.toValue(leaf.text, grammarElement.rule.name, leaf).toString()
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend:227]: `toValue()` on a leaf node could throw a value converter exception if the leaf node was used in a data type rule e.g.
`IntAsID: INT;` with `999999999999999999`.
The value converter exception is not caught / handled in `doPrepareRename`

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend (hunk 1 above)
Comment (szarnekow) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend:228]: Maybe even use `org.eclipse.xtext.linking.impl.LinkingHelper.getCrossRefNodeAsString(INode, boolean)` and find a proper cross ref node starting at the leaf node.

Comment (JanKoehnlein) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend:228]: As you wish.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend (hunk 2 above)
Comment (JanKoehnlein) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend:201]: The truth is, it doesn't even work with qualified names (as a datatype rule), unless you have registered a special value converter. My goal was to cover the basic behavior of the default ID rule. I think the method `doPrepareRename` is small enough to be overridden for more specific use cases.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend (hunk 3 above)
Comment (JanKoehnlein) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend:203]: Done_

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend (hunk 4 above)
Comment (JanKoehnlein) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend:227]: Would you really expect rename to work OOTB when integers are used as names? 
Anyway, I added some code to handle this.

**Comment by JanKoehnlein** [comment]
Thanks for the feedback.

**Comment by cdietrich** [comment]
@JanKoehnlein you missed the generated code https://github.com/eclipse/xtext-core/pull/1493

**Pull request merged**